server/src/main/java/org/opensearch/telemetry/tracing/listener/TraceableSearchRequestOperationsListener.java

```diff
@@ -0,0 +1,82 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.telemetry.tracing.listener;
+
+import org.opensearch.action.search.SearchPhaseContext;
+import org.opensearch.action.search.SearchRequestContext;
+import org.opensearch.action.search.SearchRequestOperationsListener;
+import org.opensearch.telemetry.tracing.AttributeNames;
+import org.opensearch.telemetry.tracing.Span;
+import org.opensearch.telemetry.tracing.SpanBuilder;
+import org.opensearch.telemetry.tracing.SpanContext;
+import org.opensearch.telemetry.tracing.SpanScope;
+import org.opensearch.telemetry.tracing.Tracer;
+
+import static org.opensearch.core.common.Strings.capitalize;
+
+/**
+ * SearchRequestOperationsListener subscriber for search request tracing
+ *
+ * @opensearch.internal
+ */
+public final class TraceableSearchRequestOperationsListener extends SearchRequestOperationsListener {
+    private final Tracer tracer;
+    private final Span requestSpan;
+    private Span phaseSpan;
+    private SpanScope phaseSpanScope;
+
+    public TraceableSearchRequestOperationsListener(final Tracer tracer, final Span requestSpan) {
+        this.tracer = tracer;
+        this.requestSpan = requestSpan;
+        this.phaseSpan = null;
+        this.phaseSpanScope = null;
+
+        if (tracer.isRecording() == false) {
+            this.setEnabled(false);
+        }
+    }
+
+    @Override
+    protected void onPhaseStart(SearchPhaseContext context) {
+        assert phaseSpan == null : "There should be only one search phase active at a time";
+        phaseSpan = tracer.startSpan(SpanBuilder.from(capitalize(context.getCurrentPhase().getName()), new SpanContext(requestSpan)));
+        phaseSpanScope = tracer.withSpanInScope(phaseSpan);
```

Comment:
You can just close this scope here. ref - TraceableActionListner. Same goes for endSpan.

Reply:
@dzane17 thinking about this a bit more, we have to remove the scope and keep only span (as @Gaganjuneja suggested):

```
    @Override
    protected void onPhaseStart(SearchPhaseContext context) {
        assert phaseSpan == null : "There should be only one search phase active at a time";
        phaseSpan = tracer.startSpan(SpanBuilder.from(capitalize(context.getCurrentPhase().getName()), new SpanContext(requestSpan)));
    }
```

Now, from a tracer perspective, we have a gap here, we could demarcate the span's lifetime, but not the actual work. Let's have a example here:

```java
try {
            onPhaseStart(phase);
            phase.recordAndRun();
} 

...
```

So how could we scope it? I would envision we need something like that:


```java
try {
            onPhaseStart(phase);
            try (tracer.withCurrentSpanInScope()) { /* this API does not exist yet */ 
                phase.recordAndRun();
            }
} 

...
```

The problem we have with scope is that this:
- the span will be closed somewhere in ` phase.recordAndRun();`, when `onPhaseEnd` will be called
- the scope will be very likely dealing with already closed span, this violates the scope rules (it is illogical to have scope that lives longer that the span it wraps)

@dzane17 @Gaganjuneja I think we don't have a good solution for the problem at the moment, I would suggest in this pull request that we do not have a scope but start / end span only. I will preserve timing and also casual relation (in case of nested spans).

Reply:
I have removed the scopes for now

Reply:
@reta, The problem with this is that we only add the span to the ThreadContext only during the SpanScope::attach. So even if create the span and don't attach then it will not be propagated as a parent to the other spans.

Reply:
@Gaganjuneja I totally agree with that, we only capture timing @dzane17 we cannot address this outside of this pull request - this is the whole purpose of the change. Messing  context propagation is invisible but incredibly harmful thing - it gives completely incorrect view of what is happening (we already dealt with bugs related to that), looking for the alternatives.

Reply:
Yes we should also revisit this whether  this TraceableSearchOoerationsListener is the right place to do tracing as it just gives the start and end point?

Reply:
Exactly, and it sounds to me - 99.9% it is not (but I would like to suggest the alternative, looking into that as we speak)

Reply:
@dzane17 would you mind if I push a few changes to your pull request? (I think it will be easier since there are a few places to be changed) Thank you for your patience :heart:

Reply:
@reta Sure, go ahead

Reply:
@dzane17 done, @Gaganjuneja mind please running through the suggested change in flow? tldr;
 - the phase execution is scoped within  `executePhase`, which creates the span
 - listener does not start any spans, however it captures the current span from `executePhase`
 - listener however ends the span it captured

Reply:
Thanks @reta for the change. It couldn't be cleaner than this :)